README
NAME
    Setup::Unix::User - Setup Unix user (existence, home dir, group
    memberships)

VERSION
    This document describes version 0.11 of Setup::Unix::User (from Perl
    distribution Setup-Unix-User), released on 2014-05-18.

FUNCTIONS
  add_delete_user_groups(%args) -> [status, msg, result, meta]
    Add/delete user from group memberships.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   add_to* => *array*

    *   delete_from* => *array*

    *   etc_dir => *str* (default: "/etc")

        Location of passwd files.

    *   user => *str*

        User name.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

  adduser(%args) -> [status, msg, result, meta]
    Add user.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   etc_dir => *str* (default: "/etc")

        Location of passwd files.

    *   gecos => *str*

    *   gid => *int*

        When creating group, use specific GID.

        If not specified, will search an unused GID from "min_gid" to
        "max_gid".

    *   group => *str*

        Group name.

    *   home => *str*

    *   max_gid => *int* (default: 65534)

    *   max_uid => *int* (default: 65534)

    *   min_gid => *int* (default: 1000)

    *   min_uid => *int* (default: 1000)

    *   pass => *str*

    *   shell => *str*

    *   uid => *int*

        Add with specified UID.

        If not specified, will search an unused UID from "min_uid" to
        "max_uid".

    *   user => *str*

        User name.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

  deluser(%args) -> [status, msg, result, meta]
    Delete user.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   etc_dir => *str* (default: "/etc")

        Location of passwd files.

    *   user => *str*

        User name.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

  setup_unix_user(%args) -> [status, msg, result, meta]
    Setup Unix user (existence, group memberships).

    On do, will create Unix user if not already exists. And also make sure
    user belong to specified groups (and not belong to unwanted groups).
    Return the created UID/GID in the result.

    On undo, will delete Unix user (along with its initially created home
    dir and files) if it was created by this function. Also will restore old
    group memberships.

    This function is idempotent (repeated invocations with same arguments
    has the same effect as single invocation). This function supports
    transactions.

    Arguments ('*' denotes required arguments):

    *   create_home => *bool* (default: 1)

        Whether to create home directory when creating user.

    *   etc_dir => *str* (default: "/etc")

        Location of passwd files.

    *   group => *str*

        Group name.

    *   max_gid => *any*

    *   max_new_gid => *any*

    *   max_new_uid => *any*

    *   max_uid => *any*

    *   member_of => *array*

        List of Unix group names that the user must be member of.

        If not specified, member_of will be set to just the primary group.
        The primary group will always be added even if not specified.

    *   min_gid => *any*

    *   min_new_gid => *any*

    *   min_new_uid => *any*

    *   min_uid => *any*

    *   new_gecos => *str*

    *   new_gid => *int*

        When creating group, use specific GID.

        If not specified, will search an unused GID from "min_gid" to
        "max_gid".

    *   new_home => *str*

    *   new_max_gid => *int* (default: 65534)

    *   new_max_uid => *int* (default: 65534)

    *   new_min_gid => *int* (default: 1000)

    *   new_min_uid => *int* (default: 1000)

    *   new_pass => *str*

    *   new_shell => *str*

    *   new_uid => *int*

        Add with specified UID.

        If not specified, will search an unused UID from "min_uid" to
        "max_uid".

    *   not_member_of => *array*

        List of Unix group names that the user must NOT be member of.

    *   should_already_exist => *bool*

        Whether user should already exist.

    *   should_exist => *bool* (default: 1)

        Whether user should exist.

    *   skel_dir => *str* (default: "/etc/skel")

        Directory to get skeleton files when creating user.

    *   use_skel => *bool* (default: 1)

        Whether to copy files from skeleton dir when creating user.

    *   user => *str*

        User name.

    Special arguments:

    *   -tx_action => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_action_id => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_recovery => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_rollback => *str*

        For more information on transaction, see Rinci::Transaction.

    *   -tx_v => *str*

        For more information on transaction, see Rinci::Transaction.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

FAQ
  How to create user without creating a group with the same name as that user?
    By default, "group" is set to the same name as the user. This will
    create group with the same name as the user (if the group didn't exist).
    You can set "group" to an existing group, e.g. "users" and the setup
    function will not create a new group with the same name as user. But
    note that the group must already exist (if it does not, you can create
    it first using Setup::Unix::Group).

SEE ALSO
    Setup

    Setup::Unix::Group

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Setup-Unix-User>.

SOURCE
    Source repository is at
    <https://github.com/sharyanto/perl-Setup-Unix-User>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Setup-Unix-User>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Setup-Unix-User

0.11    2014-05-17 (SHARYANTO)

        - No functional changes.

        - Replace File::Slurp with File::Slurp::Tiny.


0.10    2012-09-06 (SHARYANTO)

        [INCOMPATIBLE CHANGES]

        - Update to Rinci 1.1.28 (transaction protocol v=2).

        - Now uses Unix::Passwd::File as the backend instead of
          Passwd::Unix::Alt. This is because UPF has working locking and gives
          detailed per-call error messages instead of using global variable.

        - Parameter name change: in setup_unix_user(), 'name' becomes 'user'; in
          setup_unix_group(), 'name' becomes 'group'. This is to align better
          with Unix::Passwd::File. Some parameters also got changed.

        - Parameter default change: default 'min_new_uid' and 'min_new_gid' is
          now 1000 instead of 0. You can of course still set it to 0.


0.09    2012-03-29 (SHARYANTO)

        - Fix: Ignore membership to non-existant group.

        - Use Test::Setup.

        - build: Replace SubSpec PodWeaver plugin with Perinci.


0.08    2012-03-02 (SHARYANTO)

        - No functional changes. Replace Sub::Spec with Rinci.


0.07    2011-08-11 (SHARYANTO)

        [ENHANCEMENTS]

        - User: Add new option 'primary_group' to allow adding a user without
          creating a group with the same name as user [RT#70202]


0.06    2011-05-20 (SHARYANTO)

        - No functional changes.

        [BUG FIXES]

        - (build) Add missing dep: Setup::File.


0.05    2011-05-18 (SHARYANTO)

        - No functional changes. Tweaked log messages.


0.04    2011-05-18 (SHARYANTO)

        - No functional changes. Changed logging levels.


0.03    2011-05-17 (SHARYANTO)

        [ENHANCEMENTS]

        - setup_unix_user(): Add 'should_already_exist' parameter

        [ETC]

        - Change default max UID/GID from 65535 to 65534, to be compatible with
          Debian


0.02    2011-05-17 (SHARYANTO)

        [ENHANCEMENTS]

        - Add max_new_gid & max_new_uid parameters.

        [BUG FIXES]

        - Fixed some typos in POD

        - Bug: redo didn't recreate group with same GID.


0.01    2011-04-28 (SHARYANTO)

        - First release.


t/testlib.pl
use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use File::chdir;
use File::Path qw(remove_tree);
use File::Slurp::Tiny qw(write_file);
use File::Temp qw(tempdir);
use Setup::Unix::Group qw(setup_unix_group);
use Setup::Unix::User  qw(setup_unix_user);
use Test::More 0.96;
use Test::Perinci::Tx::Manager qw(test_tx_action);
use Unix::Passwd::File;

sub setup_data {
    unlink "$::tmp_dir/passwd";
    write_file("$::tmp_dir/passwd", <<_);
root:x:0:0:root:/root:/bin/bash
bin:x:1:1:bin:/bin:/bin/sh
daemon:x:2:2:daemon:/sbin:/bin/sh
u1:x:1000:1000::$::tmp_dir/u1:/bin/bash
u2:x:1001:1001::/home/u2:/bin/bash
_

    unlink "$::tmp_dir/shadow";
    write_file("$::tmp_dir/shadow", <<_);
root:*:14607:0:99999:7:::
bin:*:14607:0:99999:7:::
daemon:*:14607:0:99999:7:::
u1:*:14607:0:99999:7:::
u2:*:14607:0:99999:7:::
_

    unlink "$::tmp_dir/group";
    write_file("$::tmp_dir/group", <<_);
root:x:0:
bin:x:1:
daemon:x:2:
nobody:x:111:
u1:x:1000:u1
u2:x:1002:u2,u1
_

    unlink "$::tmp_dir/gshadow";
    write_file("$::tmp_dir/gshadow", <<_);
root:::
bin:::
daemon:::
nobody:!::
u1:!::
u2:!::u1
_

    # setup skeleton
    remove_tree "$::tmp_dir/skel";
    mkdir("$::tmp_dir/skel");
    mkdir("$::tmp_dir/skel/.dir1");
    write_file("$::tmp_dir/skel/.dir1/.file1", "file 1");
    write_file("$::tmp_dir/skel/.file2", "file 2");
    write_file("$::tmp_dir/skel/.file3", "file 3");

    # home dirs
    remove_tree "$::tmp_dir/home";
    remove_tree "$::tmp_dir/u1";
}

sub setup {
    plan skip_all => "No /etc/passwd, probably not Unix system"
        unless -f "/etc/passwd";

    $::tmp_dir = tempdir(CLEANUP => 1);
    $CWD = $::tmp_dir;

    setup_data();
    note "tmp dir = $::tmp_dir";
}

sub teardown {
    done_testing();
    if (Test::More->builder->is_passing) {
        #note "all tests successful, deleting temp files";
        $CWD = "/";
    } else {
        diag "there are failing tests, not deleting temp files";
    }
}

sub _test_setup_unix_group_or_user {
    my ($which, %tsuargs) = @_;

    my %ttaargs;
    for (grep {!/after_do|after_undo/} keys %tsuargs) {
        $ttaargs{$_} = $tsuargs{$_};
    }

    $ttaargs{tmpdir} = $::tmp_dir;
    $ttaargs{reset_state} = sub { setup_data() };
    $ttaargs{f} = $which eq 'group' ?
        'Setup::Unix::Group::setup_unix_group' :
            'Setup::Unix::User::setup_unix_user';
    my %fargs = %{ $tsuargs{args} };
    $fargs{etc_dir} = $::tmp_dir;
    $ttaargs{args} = \%fargs;

    for my $ak (qw/after_do after_undo/) {
        my $a = $tsuargs{$ak};
        next unless $a;
        #diag explain $a;
        $ttaargs{$ak} = sub {
            my $res;
            if ($which eq 'user') {
                $res = Unix::Passwd::File::get_user(
                    etc_dir=>$fargs{etc_dir}, user=>$fargs{user});
            } else {
                $res = Unix::Passwd::File::get_group(
                    etc_dir=>$fargs{etc_dir}, group=>$fargs{group});
            }
            #note explain $res;

            my $exists = $res->[0] == 200;

            if ($a->{exists} // 1) {
                ok($exists, "exists") or return;
                my $e = $res->[2];
                if ($which eq 'user') {
                    for (qw/uid gid gecos home shell/) {
                        if (defined $a->{$_}) {
                            is($e->{$_}, $a->{$_}, $_);
                        }
                    }
                } else {
                    if (defined $a->{gid}) {
                        is($e->{gid}, $a->{gid}, "gid");
                    }
                }
            } else {
                ok(!$exists, "does not exist");
            }

            if ($which eq 'user') {
                if ($a->{member_of}) {
                    for my $g (@{ $a->{member_of} }) {
                        $res = Unix::Passwd::File::is_member(
                            etc_dir=>$fargs{etc_dir},
                            user=>$fargs{user}, group=>$g);
                        ok($res,
                           "user $fargs{user} is member of $g");
                    }
                }
                if ($a->{not_member_of}) {
                    for my $g (@{ $a->{not_member_of} }) {
                        $res = Unix::Passwd::File::is_member(
                            etc_dir=>$fargs{etc_dir},
                            user=>$fargs{user}, group=>$g);
                        ok(defined($res) && !$res,
                           "user $fargs{user} is not member of $g");
                    }
                }
            }

            if ($a->{extra}) {
                $a->{extra}->();
            }
        };
    }

    test_tx_action(%ttaargs);
}

sub test_setup_unix_group { _test_setup_unix_group_or_user('group', @_) }

sub test_setup_unix_user  { _test_setup_unix_group_or_user('user',  @_) }

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Crypt::Password::Util" => 0,
    "File::Copy::Recursive" => 0,
    "File::Slurp" => 0,
    "File::chdir" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.98",
    "Test::Perinci::Tx::Manager" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Setup Unix user (existence, home dir, group memberships)",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Setup-Unix-User",
  "dist_version" => "0.10",
  "license" => "perl",
  "module_name" => "Setup::Unix::User",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::Copy::Undoable" => "0.02",
    "File::Trash::Undoable" => 0,
    "Log::Any" => 0,
    "PerlX::Maybe" => 0,
    "Rinci" => "v1.1.28",
    "Setup" => "v1.0.3",
    "Setup::File" => 0,
    "Text::Password::Pronounceable" => 0,
    "Unix::Passwd::File" => "0.07",
    "perl" => "5.010001"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Setup/Unix/Group.pm
package Setup::Unix::Group;

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(setup_unix_group);

use PerlX::Maybe;
use Unix::Passwd::File;

our $VERSION = '0.11'; # VERSION

our %SPEC;

my %common_args = (
    etc_dir => {
        summary => 'Location of passwd files',
        schema  => ['str*' => {default=>'/etc'}],
    },
    group => {
        schema  => 'str*',
        summary => 'Group name',
    },
);

$SPEC{delgroup} = {
    v => 1.1,
    summary => 'Delete group',
    description => <<'_',

Fixed state: group does not exist.

Fixable state: group exists.

_
    args => {
        %common_args,
    },
    features => {
        tx => {v=>2},
        idempotent => 1,
    },
};
sub delgroup {
    my %args = @_;

    my $tx_action = $args{-tx_action} // '';
    my $dry_run   = $args{-tx_action} // '';
    my $group     = $args{group} or return [400, "Please specify group"];
    $group =~ $Unix::Passwd::File::re_group
        or return [400, "Invalid group"];
    my %ca        = (etc_dir => $args{etc_dir}, group=>$group);
    my $res;

    if ($tx_action eq 'check_state') {
        my $res = Unix::Passwd::File::get_group(%ca);
        return $res unless $res->[0] == 200 || $res->[0] == 404;

        return [304, "Group $group already doesn't exist"] if $res->[0] == 404;
        $log->info("(DRY) Deleting Unix group $group ...") if $dry_run;
        return [200, "Group $group needs to be deleted", undef, {undo_actions=>[
            [addgroup => {%ca, gid => $res->[2]{gid}}],
        ]}];
    } elsif ($tx_action eq 'fix_state') {
        # we don't want to have to get_group() when fixing state, to reduce
        # number of read passes to the passwd files
        $log->info("Deleting Unix group $group ...");
        return Unix::Passwd::File::delete_group(%ca);
    }
    [400, "Invalid -tx_action"];
}

$SPEC{addgroup} = {
    v => 1.1,
    summary => 'Add group',
    args => {
        %common_args,
        gid => {
            summary => 'Add with specified GID',
            description => <<'_',

If not specified, will search an unused GID from `min_new_gid` to `max_new_gid`.

If specified, will accept non-unique GID (that which has been used by other
group).

_
            schema => 'int',
        },
        min_gid => {
            summary => 'Specify range for new GID',
            description => <<'_',

If a free GID between `min_gid` and `max_gid` is not available, an error is
returned.

Passed to Unix::Passwd::File's `min_new_gid`.

_
            schema => [int => {between=>[0, 65535], default=>1000}],
        },
        max_gid => {
            summary => 'Specify range for new GID',
            description => <<'_',

If a free GID between `min_gid` and `max_gid` is not available, an error is
returned.

Passed to Unix::Passwd::File's `max_new_gid`.

_
            schema => [int => {between=>[0, 65535], default=>65534}],
        },
    },
    features => {
        tx => {v=>2},
        idempotent => 1,
    },
};
sub addgroup {
    my %args = @_;

    my $tx_action = $args{-tx_action} // '';
    my $dry_run   = $args{-dry_run};
    my $group     = $args{group} or return [400, "Please specify group"];
    $group =~ $Unix::Passwd::File::re_group
        or return [400, "Invalid group"];
    my $gid       = $args{gid};
    my $min_gid   = $args{min_gid} //  1000;
    my $max_gid   = $args{max_gid} // 65534;
    my %ca0       = (etc_dir => $args{etc_dir});
    my %ca        = (%ca0, group=>$group);
    my $res;

    if ($tx_action eq 'check_state') {
        $res = Unix::Passwd::File::get_group(%ca);
        return $res unless $res->[0] == 200 || $res->[0] == 404;

        if ($res->[0] == 200) {
            if (!defined($gid) || $gid == $res->[2]{gid}) {
                return [304, "Group $group already exists"];
            } else {
                return [412, "Group $group already exists but with different ".
                            "GID ($res->[2]{gid}, wanted $gid)"];
            }
        } else {
            $log->info("(DRY) Adding Unix group $group ...") if $dry_run;
            return [200, "Group $group needs to be added", undef,
                    {undo_actions=>[
                        [delgroup => {%ca}],
            ]}];
        }
    } elsif ($tx_action eq 'fix_state') {
        # we don't want to have to get_group() when fixing state, to reduce
        # number of read passes to the passwd files
        $log->info("Adding Unix group $group ...");
        $res = Unix::Passwd::File::add_group(
            %ca,
            maybe gid     => $gid,
            min_gid => $min_gid,
            max_gid => $max_gid);
        if ($res->[0] == 200) {
            $args{-stash}{result}{gid} = $res->[2]{gid};
            return [200, "Created"];
        } else {
            return $res;
        }
    }
    [400, "Invalid -tx_action"];
}

$SPEC{setup_unix_group} = {
    v           => 1.1,
    summary     => "Setup Unix group (existence)",
    description => <<'_',

On do, will create Unix group if not already exists. The created GID will be
returned in the result (`{gid => GID}`). If `should_already_exist` is set to
true, won't create but only require that group already exists. If `should_exist`
is set to false, will delete existing group instead of creating it.

On undo, will delete Unix group previously created.

On redo, will recreate the Unix group with the same GID.

_
    args => {
        should_exist => {
            summary => 'Whether group should exist',
            schema  => [bool => {default=>1}],
        },
        should_already_exist => {
            summary => 'Whether group should already exist',
            schema  => 'bool',
        },
        %{ $SPEC{addgroup}{args} },
    },
    features => {
        tx => {v=>2},
        idempotent => 1,
    },
};
for (qw/setup_unix_group/) {
    $SPEC{$_}{args}{min_new_gid} = delete $SPEC{$_}{args}{min_gid};
    $SPEC{$_}{args}{max_new_gid} = delete $SPEC{$_}{args}{max_gid};
    $SPEC{$_}{args}{new_gid}     = delete $SPEC{$_}{args}{gid};
}
sub setup_unix_group {
    my %args = @_;

    # TMP, schema
    my $dry_run       = $args{-dry_run};
    my $group         = $args{group} or return [400, "Please specify group"];
    $group =~ $Unix::Passwd::File::re_group
        or return [400, "Invalid group"];
    my $should_exist  = $args{should_exist} // 1;
    my $should_aexist = $args{should_already_exist};
    my %ca            = (etc_dir=>$args{etc_dir}, group=>$group);

    my $exists = Unix::Passwd::File::group_exists(%ca);
    my (@do, @undo);

    #$log->tracef("group=%s, exists=%s, should_exist=%s, ", $group, $exists, $should_exist);
    if ($exists) {
        if (!$should_exist) {
            $log->info("(DRY) Deleting group $group ...");
            push    @do  , [delgroup=>{%ca}];
            unshift @undo, [addgroup=>{
                %ca,
                maybe gid     => $args{new_gid},
                maybe min_gid => $args{min_new_gid},
                maybe max_gid => $args{max_new_gid},
            }];
        }
    } else {
        if ($should_aexist) {
            return [412, "Group $group should already exist"];
        } elsif ($should_exist) {
            $log->info("(DRY) Adding group $group ...");
            push    @do  , [addgroup=>{
                %ca,
                maybe gid     => $args{new_gid},
                maybe min_gid => $args{min_new_gid},
                maybe max_gid => $args{max_new_gid},
            }];
            unshift @do  , [delgroup=>{%ca}];
        }
    }

    if (@do) {
        return [200, "", undef, {do_actions=>\@do, undo_actions=>\@undo}];
    } else {
        return [304, "Already fixed"];
    }
}

1;
# ABSTRACT: Setup Unix group (existence)

__END__

=pod

=encoding UTF-8

=head1 NAME

Setup::Unix::Group - Setup Unix group (existence)

=head1 VERSION

This document describes version 0.11 of Setup::Unix::Group (from Perl distribution Setup-Unix-User), released on 2014-05-18.

=head1 FUNCTIONS


=head2 addgroup(%args) -> [status, msg, result, meta]

Add group.

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<etc_dir> => I<str> (default: "/etc")

Location of passwd files.

=item * B<gid> => I<int>

Add with specified GID.

If not specified, will search an unused GID from C<min_new_gid> to C<max_new_gid>.

If specified, will accept non-unique GID (that which has been used by other
group).

=item * B<group> => I<str>

Group name.

=item * B<max_gid> => I<int> (default: 65534)

Specify range for new GID.

If a free GID between C<min_gid> and C<max_gid> is not available, an error is
returned.

Passed to Unix::Passwd::File's C<max_new_gid>.

=item * B<min_gid> => I<int> (default: 1000)

Specify range for new GID.

If a free GID between C<min_gid> and C<max_gid> is not available, an error is
returned.

Passed to Unix::Passwd::File's C<min_new_gid>.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.


=head2 delgroup(%args) -> [status, msg, result, meta]

Delete group.

Fixed state: group does not exist.

Fixable state: group exists.

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<etc_dir> => I<str> (default: "/etc")

Location of passwd files.

=item * B<group> => I<str>

Group name.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.


=head2 setup_unix_group(%args) -> [status, msg, result, meta]

Setup Unix group (existence).

On do, will create Unix group if not already exists. The created GID will be
returned in the result (C<{gid => GID}>). If C<should_already_exist> is set to
true, won't create but only require that group already exists. If C<should_exist>
is set to false, will delete existing group instead of creating it.

On undo, will delete Unix group previously created.

On redo, will recreate the Unix group with the same GID.

This function is idempotent (repeated invocations with same arguments has the same effect as single invocation). This function supports transactions.


Arguments ('*' denotes required arguments):

=over 4

=item * B<etc_dir> => I<str> (default: "/etc")

Location of passwd files.

=item * B<group> => I<str>

Group name.

=item * B<max_new_gid> => I<int> (default: 65534)

Specify range for new GID.

If a free GID between C<min_gid> and C<max_gid> is not available, an error is
returned.

Passed to Unix::Passwd::File's C<max_new_gid>.

=item * B<min_new_gid> => I<int> (default: 1000)

Specify range for new GID.

If a free GID between C<min_gid> and C<max_gid> is not available, an error is
returned.

Passed to Unix::Passwd::File's C<min_new_gid>.

=item * B<new_gid> => I<int>

Add with specified GID.

If not specified, will search an unused GID from C<min_new_gid> to C<max_new_gid>.

If specified, will accept non-unique GID (that which has been used by other
group).

=item * B<should_already_exist> => I<bool>

Whether group should already exist.

=item * B<should_exist> => I<bool> (default: 1)

Whether group should exist.

=back

Special arguments:

=over 4

=item * B<-tx_action> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_action_id> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_recovery> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_rollback> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=item * B<-tx_v> => I<str>

For more information on transaction, see L<Rinci::Transaction>.

=back

Return value:

Returns an enveloped result (an array).

First element (status) is an integer containing HTTP status code
(200 means OK, 4xx caller error, 5xx function error). Second element
(msg) is a string containing error message, or 'OK' if status is
200. Third element (result) is optional, the actual result. Fourth
element (meta) is called result metadata and is optional, a hash
that contains extra information.

=head1 FAQ

=head1 SEE ALSO

L<Setup>

L<Setup::Unix::User>

=head1 HOMEPAGE

Please visit the project's homepage at L<https://metacpan.org/release/Setup-Unix-User>.

=head1 SOURCE

Source repository is at L<https://github.com/sharyanto/perl-Setup-Unix-User>.

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website L<https://rt.cpan.org/Public/Dist/Display.html?Name=Setup-Unix-User>

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Setup Unix user (existence, home dir, group memberships)",
  "AUTHOR" => "Steven Haryanto <stevenharyanto\@gmail.com>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Setup-Unix-User",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Setup::Unix::User",
  "PREREQ_PM" => {
    "File::Copy::Undoable" => "0.02",
    "File::Trash::Undoable" => 0,
    "Log::Any" => 0,
    "PerlX::Maybe" => 0,
    "Rinci" => "v1.1.28",
    "Setup" => "v1.0.3",
    "Setup::File" => 0,
    "Text::Password::Pronounceable" => 0,
    "Unix::Passwd::File" => "0.07",
    "experimental" => 0
  },
  "TEST_REQUIRES" => {
    "Crypt::Password::Util" => 0,
    "File::Copy::Recursive" => 0,
    "File::Slurp::Tiny" => 0,
    "File::Spec" => 0,
    "File::chdir" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98",
    "Test::Perinci::Tx::Manager" => 0
  },
  "VERSION" => "0.11",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Crypt::Password::Util" => 0,
  "File::Copy::Recursive" => 0,
  "File::Copy::Undoable" => "0.02",
  "File::Slurp::Tiny" => 0,
  "File::Spec" => 0,
  "File::Trash::Undoable" => 0,
  "File::chdir" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Log::Any" => 0,
  "PerlX::Maybe" => 0,
  "Rinci" => "v1.1.28",
  "Setup" => "v1.0.3",
  "Setup::File" => 0,
  "Test::More" => "0.98",
  "Test::Perinci::Tx::Manager" => 0,
  "Text::Password::Pronounceable" => 0,
  "Unix::Passwd::File" => "0.07",
  "experimental" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



